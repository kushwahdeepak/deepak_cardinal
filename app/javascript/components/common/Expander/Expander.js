import React, { useState } from 'react'
import { Accordion } from 'react-bootstrap'
import Card from 'react-bootstrap/Card'
import Button from 'react-bootstrap/Button'
import { useAccordionToggle } from 'react-bootstrap/AccordionToggle'

import './style/expander.scss'
import { Head, Icon } from './style/Expander.styled'

function Expander({ label, expanded, children, candidatePage }) {
    return (
        <Accordion
        className="text-font"
            data-test="accordion-item"
            defaultActiveKey={expanded ? '0' : '-1'}
            style={{width: "210px"}}
        >
            <Card style={{ border: 'none' }}>
                <CustomToggle eventKey="0" candidatePage={candidatePage}>{label}</CustomToggle>
                <Accordion.Collapse eventKey="0">
                    <Card.Body style={{padding:'0px 1.25rem '}}>
                        {children}
                    </Card.Body>
                </Accordion.Collapse>
            </Card>
        </Accordion>
    )
}

function CustomToggle({ children, eventKey, candidatePage }) {
    const [expanded, setExpanded] = useState(true)

    const decoratedOnClick = useAccordionToggle(eventKey, () => {
        setExpanded((expanded) => !expanded)
    })

    return (
        <Button
            type="Button"
            variant="expander"
            style={{
                border: 'none',
                fontWeight: 500,
                backgroundColor: candidatePage ? 'white' : '',
            }}
            className= {candidatePage && expanded ?
                'candidate-class' : candidatePage && !expanded ?
                 'candidate-class expended-color-active' : ''}
            onClick={decoratedOnClick}
            data-test="expander-button"
        >
            <Head>
                <label style={{ fontSize: '13px' }}>{children}</label>
                <Icon direction={expanded ? 1 : 0}>
                    <i data-feather="chevron-down" />
                </Icon>
            </Head>
        </Button>
    )
}

export default Expander
