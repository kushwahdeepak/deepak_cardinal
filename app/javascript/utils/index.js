import { act } from 'react-dom/test-utils'

export const findByTestAttr = (wrapper, attr) => {
    return wrapper.find(`[data-test='${attr}']`)
}

export const timeoutResolve = (time = 0, response = {}) =>
    new Promise((resolve) => {
        setTimeout(() => resolve(response), time)
    })

export const wait = (time = 0) => {
    const timer = timeoutResolve(time)

    return act(async () => {
        await timer
    })
}
export function emailIsValid(email) {
    return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email)
}

export function generateRandom(min, max) {
    return Math.floor(min + Math.random() * (max + 1 - min))
}

export function capitalizeFirstLetter(string) {
    return string.charAt(0).toUpperCase() + string.slice(1).toLowerCase()
}

export function createMarkedList(array) {
    if (!Array.isArray(array)) {
        return []
    }
    return array.map((candidate) => ({
        id: candidate.id,
        name: candidate.id,
        fullName: `${candidate.first_name} ${candidate.last_name}`,
        emailAddress: candidate.email_address,
        isChecked: false,
    }))
}

export function stringReplace(string,toSymbol,fromSymbol){
    return string.replace(toSymbol, fromSymbol)
}

export function capitalize(string) {
    return string.replace(/\b(\w)/g, s => s.toUpperCase());
}

export function skillIsNotEmpty(value){
    return value.trim().length > 0;
}

export function firstCharacterCapital(value){
    return value?.charAt(0)?.toUpperCase() + value?.slice(1)
}
