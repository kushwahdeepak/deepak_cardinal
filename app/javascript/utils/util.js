import isNil from 'lodash.isnil'
import isEmpty from 'lodash.isempty'

class Util {
    /**
     * Define stage values
     */
    static STAGES = ['first_interview', 'second_interview', 'offer', 'reject']

    /**
     * Tests whether a single object attribute includes the at
     * least one string from the given set.
     */
    static attributePassesStringSetFilter(object, attr, stringSet) {
        if (isNil(object[attr])) return false

        // We "OR" the tag tests: the attribute needs to contain only one tag
        return stringSet
            .map((string) => string.trim().toLowerCase())
            .reduce(
                (passes, string) =>
                    passes ||
                    Util.attributeToString(object[attr])
                        .toLowerCase()
                        .includes(string),
                false
            )
    }

    /**
     * Tests whether at least one of a set of object attributes
     * includes every word from a given piece of text.
     */
    static objectPassesTextFilter(object, attrs, text) {
        text = text.trim().toLowerCase()

        if (text === '') return true

        const words = text.split(' ')

        // Join all the given attribute names to lowercase, stringified attribute values
        const allAttrsString = attrs
            .filter((attr) => !isNil(object[attr]))
            .map((attr) => Util.attributeToString(object[attr]).toLowerCase())
            .join('')

        // Is every word included in the allAttrsString?
        return words.reduce(
            (matchesAll, word) => matchesAll && allAttrsString.includes(word),
            true
        )
    }

    static attributeToString(attr) {
        if (typeof attr === 'string') {
            return attr
        } else if (Array.isArray(attr)) {
            return attr.join(',')
        } else if (typeof attr === 'object') {
            return JSON.stringify(attr)
        } else {
            return '' + attr
        }
    }

    static numberWithCommas(x) {
        return x?.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',')
    }

    /**
     * Returns number of results in format [x] - [y] of [z]
     * @param {number} totalJobsCount Total number of jobs
     * @param {number} pageCount Total number of pages
     * @param {number} activePage Currently active page
     * @param {number} pageSize How many results to display per page
     */
    static displayNumberOfResults(
        totalJobsCount,
        pageCount,
        activePage,
        pageSize,
        totalPersonCount=null,
        candidatePage=false,
        showdisplaycount=true
    ) {
        if (totalJobsCount == 0) return '0 - 0 of 0'
        const offset = pageSize * pageCount - totalJobsCount
        const lastJobIndex =
            (activePage + 1) * pageSize > totalJobsCount
                ? totalJobsCount
                : (activePage + 1) * pageSize
        const firstJobIndex =
            activePage == 0
                ? 1
                : activePage == pageCount - 1
                ? lastJobIndex - pageSize + offset + 1
                : lastJobIndex - pageSize + 1
        return `${firstJobIndex} - ${lastJobIndex} of ${Util.numberWithCommas(
            (candidatePage && showdisplaycount) ? totalPersonCount : totalJobsCount

        )}`
    }

    /**
     * Returns a string indicating the candidate's title and current
     * company, based on available information on the Person object
     * @param {object} candidate Person object from our db
     */
    static candidateCompanyString(candidate) {
        const position = candidate.company_position
        const currentCompany = candidate.current_company

        if (isEmpty(position) && isEmpty(currentCompany)) return ''

        if (!isEmpty(position) && !isEmpty(currentCompany))
            return `${position} at ${currentCompany}`

        if (!isEmpty(position)) return position

        if (!isEmpty(currentCompany)) return `at ${currentCompany}`
    }

    /**
     * Converts the given array of objects with ids to an object whose
     * keys are the ids of the array objects.
     *
     * Every object in the given array must have a unique 'id' property defined.
     *
     * @param {Array} array
     * @returns {Object}
     */
    static arrayToObject(array) {
        const obj = {}
        array.forEach((element) => (obj[element.id] = element))
        return obj
    }

    /**
     * Makes a copy of the given array and replace the value at 'index' with 'value'
     *
     * @param {Array} array
     * @param {Number} index
     * @param {*} value
     * @returns {Array}
     */
    static replaceInArrayCopy_i(array, index, value) {
        const clone = array.slice()
        clone[index] = value
        return clone
    }

    /**
     * Makes a copy of the given array and replaces the element with the same
     * .id property as obj.id with obj
     *
     * @param {Array} array
     * @param {Object} obj
     * @returns {Array}
     */
    static replaceInArrayCopy(array, obj) {
        const index = array.findIndex((element) => element.id === obj.id)
        return Util.replaceInArrayCopy_i(array, index, obj)
    }

    static deleteInArrayCopy(array, index) {
        const copy = array.slice()
        copy.splice(index, 1)

        return copy
    }

    static handleUndefined(value){
        if (value === undefined || value === 'undefined')
        return '';
        else
        return value
    }

    static handleUndefinedFullName(firstName, lastName){
      if(firstName === "undefined" || firstName === undefined)
      return lastName
      else if(lastName === "undefined" || lastName === undefined)
      return firstName
      else
      return firstName + ' ' + lastName     }

    static capitalize(value){
         return value.charAt(0).toUpperCase() + value.slice(1).toLowerCase();
    }
}

export default Util
