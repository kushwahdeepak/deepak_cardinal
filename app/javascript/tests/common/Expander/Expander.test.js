import React from 'react';
import Expander from '../../../components/common/Expander/Expander';
import renderer from 'react-test-renderer';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findByTestAttr } from '../../../utils';
import { act } from 'react-dom/test-utils';

configure({adapter: new Adapter()});

const initialShallowRender = (props = {}) => {
	return shallow(<Expander {...props}/>);
};

const initialMountRender = (props = {}) => {
	return mount(<Expander {...props}/>);
};

const initialProps = {
};

describe('Expander component',  () => {
	let wrapper;
	let fullWrapper;

	beforeEach(() => {

		const props = {
			...initialProps
		};

		wrapper = initialShallowRender(props);
		fullWrapper = initialMountRender(props);
	});

	afterEach(() => {
		fullWrapper && fullWrapper.unmount();
	});

	it('should render Expander component', () => {
		expect(wrapper.exists()).toBe(true);
	});

	it('should match with a snapshot', () => {
		const component = renderer.create(<Expander />);
		let tree = component.toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('Expander component with props', () => {
		let wrapper;
		let fullWrapper;
		const setState = jest.fn();
		const useStateSpy = jest.spyOn(React, 'useState');
		useStateSpy.mockImplementation((init) => [init, setState]);
		const actions = async (wrapper, _actions) => {
			await act(async () => {
				await (new Promise(resolve => setTimeout(resolve, 0)));
				_actions();
				wrapper.update();
			});
		};

		beforeEach(() => {
			const props = {
				...initialProps,
				expanded: true
			};

			wrapper = initialShallowRender(props);
			fullWrapper = initialMountRender(props);
		});

		afterEach(() => {
			fullWrapper && fullWrapper.unmount();
		});

		it('should render correct defaultActiveKey prop', () => {
			const component = findByTestAttr(wrapper, 'accordion-item');
			expect(component.prop('defaultActiveKey')).toBe('0');
		});
		it('should call onClick handler', async () => {
			const component = findByTestAttr(fullWrapper, 'expander-button').at(0);
			await actions(component, () => {
				component.props().onClick();
			});
			expect(setState).toHaveBeenCalledTimes(1);
		});
	});
});
