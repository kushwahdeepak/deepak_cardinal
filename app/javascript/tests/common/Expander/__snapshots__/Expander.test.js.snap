// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expander component should match with a snapshot 1`] = `
<div
  className="accordion"
  data-test="accordion-item"
>
  <button
    className="btn btn-expander"
    data-test="expander-button"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "border": "none",
        "fontWeight": 500,
        "width": "100%",
      }
    }
    type="Button"
  >
    <div
      style={
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <div
        style={
          Object {
            "display": "block",
          }
        }
      >
        <i
          data-feather="chevron-down"
        />
      </div>
      <div
        style={
          Object {
            "display": "none",
          }
        }
      >
        <i
          data-feather="chevron-right"
        />
      </div>
    </div>
  </button>
  <div
    aria-expanded={null}
    className="collapse"
  >
    <div
      className="container"
      style={
        Object {
          "paddingBottom": "0",
        }
      }
    />
  </div>
</div>
`;
